---
title: "Differential Gene Expression Analysis of hornet castes"
author: "Emeline Favreau"
date: '2023-01-31'
output:
  pdf_document: default
  html_document: default
---

#### Copyright 2023 Emeline Favreau, University College London.

This is a short version of code for our Scientific Reports manuscript.
For the longer version, check it on 
[Github](https://github.com/EmelineFavreau/Vespa-Genomes-Analyses).

### Objectives of analysis

Brains of workers (5 pools), gynes (5 pools) and 1 queen (no pool) of
*Vespa crabro* have been sequenced for RNA. 
The behavioural differences between castes are expected to be measured using
brain RNAseq as proxy. 
Which genes are significantly differentially expressed between castes?

### Analysis steps:

-   Obtaining data
-   Formatting data
-   Aim 1: Run DESeq2 on all samples




```{r load all the libraries, eval = TRUE, echo = FALSE, include = FALSE}
# list libraries
basic_libraries <- c("ggplot2",
                     "tidyverse",
                     "DESeq2",
                     "apeglm",
                     "pheatmap",
                     "RColorBrewer",
                     "readr")

# load libraries
for (lib in basic_libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                install.packages(lib)
                library(lib, character.only = TRUE)
        }
}
```

```{r import data, eval = TRUE, echo = FALSE, include = FALSE}
# add all the files needed here, with descriptions

# matrix of read counts from Nextflow QC pipeline
# including gene ID as row
# samples as columns
# queens: VC, VCQ2
# gynes: VCG1, VCG2, VCG3, VCG4, VCG5
# workers: VC_W1, VC_W5, VC_W6, VC_W7
cts <- as.matrix(read.csv(
  "input/merged_gene_counts.txt",
  sep = "\t",
  row.names = "Geneid"))
```

```{r formatting data, eval = TRUE, echo = FALSE, include = TRUE}
# for this analysis, we remove the queens
cts1 <- cts[, !colnames(cts) %in% c("VC_1Aligned.sortedByCoord.out.bam",
                                    "VCQ2_1Aligned.sortedByCoord.out.bam")]


# remove gene_name (empty field)
cts2 <- cts1[, !colnames(cts1) %in% c("gene_name")]

# make a phenotype data
# sample, reproductives, castes
# reproductives = gynes (i.e. a non-mated queen) or workers
# castes = workers or not
pheno_df <- data.frame(colnames(cts2),
                       colnames(cts2),
                       row.names = colnames(cts2),
                       stringsAsFactors = FALSE)

# name the columns
colnames(pheno_df) <- c("reproduction", "caste")



# categorise samples by reproduction
pheno_df$reproduction[grep(pattern = "VC_W",
                           x = colnames(cts2),
                           value = FALSE)] <- "nonReproductive"

pheno_df$reproduction[grep(pattern = "VC_W",
                           x = colnames(cts2),
                           value = FALSE,
                      invert = TRUE)] <- "reproductive"


# categorise samples by caste
pheno_df$caste[grep(pattern = "VC_W",
                           x = colnames(cts2),
                           value = FALSE)] <- "worker"

pheno_df$caste[grep(pattern = "VCG",
                           x = colnames(cts2),
                           value = FALSE)] <- "gyne"




# change the names for biological meaningful
rownames(pheno_df) <- c("gyne4",
                       "worker5",
                       "gyne1",
                       "worker7",
                       "gyne2",
                       "gyne5",
                       "worker6",
                       "gyne3",
                       "worker1")

# name columns
colnames(cts2) <- rownames(pheno_df)

# change class for factor
pheno_df$reproduction <- factor(pheno_df$reproduction)
pheno_df$caste        <- factor(pheno_df$caste)
```

```{r create function for more axes, eval = TRUE, echo = FALSE, include = TRUE}
# function to build more PCA axis comparison
plot_pca_more_axes <- function(object,
                            intgroup = "condition",
                            ntop = 500,
                            x_axis = 1,
                            y_axis = 2,
                            return_data = FALSE) {
  # calculate the variance for each gene
  rv <- rowVars(assay(object))

  # select the ntop genes by variance
  select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, length(rv)))]

  # perform a PCA on the data in assay(x) for the selected genes
  pca <- prcomp(t(assay(object)[select, ]))

  # the contribution to the total variance for each component
  percent_var <- pca$sdev^2 / sum(pca$sdev^2)

  # stop plotting if missing info
  if (!all(intgroup %in% names(colData(object)))) {
    stop("the argument 'intgroup' should specify columns of colData(dds)")
  }

  # make an ingroup df
  intgroup_df <- as.data.frame(colData(object)[, intgroup, drop = FALSE])

  # add the intgroup factors together to create a new grouping factor
  group <- if (length(intgroup) > 1) {
    factor(apply(intgroup_df, 1, paste, collapse = ":"))
  } else {
    colData(object)[[intgroup]]
  }

  # assembly the data for the plot
  d <- data.frame(x_axis_for_pca = pca$x[, x_axis],
                  y_axis_for_pca = pca$x[, y_axis],
                  group = group,
                  intgroup_df,
                  name = colnames(object))

  # include percentage variable if mentioned
  if (return_data) {
    attr(d, "percent_var") <- percent_var[x_axis:y_axis]
    return(d)
  }

  # create name of axis based on the assigned value
  my_x_axis_name <- paste("PC", x_axis, sep = "_")
  my_y_axis_name <- paste("PC", y_axis, sep = "_")

  # mention PC in the axis name
  x_pc <- paste("PC", x_axis, sep = "")
  y_pc <- paste("PC", y_axis, sep = "")

  # plot the PCA
  ggplot(data = d,
         aes_string(x = d$x_axis_for_pca,
                    y = d$y_axis_for_pca,
                    color = "group")) +

          geom_point(size = 3) +

    xlab(paste0(my_x_axis_name, ": ",
                round(percent_var[x_axis] * 100),
                "% variance")) +

      ylab(paste0(my_y_axis_name, ": ",
                  round(percent_var[y_axis] * 100),
                  "% variance")) +
        coord_fixed()
}

# this could be useful if one wants to plot further axes, e.g. PC2 and PC3
```

### Aim 1: Run DESeq2 on all samples

I first check potential sample outliers looking at normalised read counts 
(with a heatmap, a sample-to-sample dendogram, a PCA) 
and a barplot of DEG value ranges.

```{r aim 1 data exploration, eval = TRUE, echo = FALSE, include = TRUE}
# remove genes for which all samples have zero counts
# because it is not biological meaningful in our analysis
cts3 <- cts2[(rowSums(cts2) != 0), ]

# make a DESeq dataset
# measure the effect of reproduction
dds <- DESeqDataSetFromMatrix(countData = cts3,
                              colData   = pheno_df,
                              design    = ~ reproduction)

# run the analysis
dds <- DESeq(dds)

# data to explore
my_data <- dds

# Read counts after data transformation
# obtain data for the first 20 genes
select <- order(rowMeans(counts(my_data,
                                normalized = TRUE)),
                decreasing = TRUE)[1:20]

# make a dataframe for reproduction
df <- as.data.frame(colData(my_data)[, c("reproduction")])

# name the column and rows
colnames(df) <- "castes"


# create a DESeqTransform object
# this gives log2(n + 1), the most basic transformation
ntd <- normTransform(my_data)

# name the row names
rownames(df) <- colnames(assay(ntd)[select, ])

# plot a heat map
pheatmap(assay(ntd)[select, ],
         cluster_rows    = FALSE,
         show_rownames   = FALSE,
         cluster_cols    = FALSE,
         annotation_col  = df,
         legend_labels   = "Caste",
         main            = "Counts of 20 top genes after log2 + 1")



# explore sample-tp-sample distances
sample_dists <- dist(t(assay(ntd)))

# transform as a matrix
sample_dist_matrix <- as.matrix(sample_dists)

# rownames are caste
rownames(sample_dist_matrix) <- ntd$caste

# column names are set to NULL
colnames(sample_dist_matrix) <- NULL

# pick a nice colour palette in a vector
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)

# plot a heatmap
pheatmap(sample_dist_matrix,
         clustering_distance_rows = sample_dists,
         clustering_distance_cols = sample_dists,
         col = colors,
         main = "Sample-to-sample distances")



## set the PCs
x_axis <- 1
y_axis <- 2

# run the plot
plot_pca_more_axes(ntd,
                        intgroup = c("caste"),
                        x_axis = x_axis,
                        y_axis = y_axis)


# check for sample outlier
boxplot(log10(assays(dds)[["cooks"]]),
         range = 0,
         las = 2)

```



```{r aim 1 explore deseq results, eval = TRUE, echo = FALSE, include = TRUE}
# extract a results table with log2 fold changes, p values (nonadj and adj)
res <- results(dds,
               name = "reproduction_reproductive_vs_nonReproductive")

# p-values exploration
res_ordered <- res[order(res$pvalue), ]

# change results for 0.05 cut-off
res05 <- results(dds, alpha = 0.05)

# Explore results
plotMA(res05, ylim = c(-10, 8))

# filter data for those genes with more than log fold change of 2
refined_results <- subset(res05,
                          (abs(log2FoldChange) > 2))


# remove the genes that have a adjust p-value of NA
# NA is from DESEQ2, assigned either because zero counts for all samples,
# or one sample being outlier
refined_no_na_results <- refined_results[!is.na(refined_results$padj), ]

# investigate these na-pvalues
# (are they related to the sole queen or sole male?)
refined_is_na_results <- refined_results[is.na(refined_results$padj), ]

# check if all samples have zero counts
one_sample_outlier_count <- 0
all_read_counts_are_zero_count <- 0
for (gene in row.names(refined_is_na_results)){

  if (mean(cts[grep(gene, x = row.names(cts)), 2:11]) > 0) {
    one_sample_outlier_count <- one_sample_outlier_count + 1
  } else {
    all_read_counts_are_zero_count <- all_read_counts_are_zero_count + 1
    }

}

# check results - sanity check - uncomment if needed
#print(paste("Number of cases with one single outlier: ",
#            one_sample_outlier_count))
#print(paste("Number of cases with all read counts equal 0: ",
#            all_read_counts_are_zero_count))


```


I find 76 downregulated genes and 64 upregulated genes in gynes.
I can now proceed to GO terms descriptions.

```{r save results, eval = TRUE, echo = FALSE, include = TRUE}
# split the DEG list into up and downregulated gene lists (in reproductives)
# so we can see in what way these genes are present in the DEG list
# 76 downregulated
# e.g. Vcabro1a004104
downregulated_gene_results <-
  refined_no_na_results[refined_no_na_results$log2FoldChange < 0, ]

# 64 upregulated
# e.g. Vcabro1a014716
upregulated_gene_results <-
  refined_no_na_results[refined_no_na_results$log2FoldChange > 0, ]

# save these tables
write.table(x = downregulated_gene_results,
          file = "result/downregulated_gene_results",
          sep = "\t",
          quote = FALSE)

write.table(x = upregulated_gene_results,
          file = "result/upregulated_gene_results",
          sep = "\t",
          quote = FALSE)
```

```{r record versions of R, eval = TRUE, echo = FALSE, include = TRUE}
# record versions of R and packages here
# sessionInfo()
#R version 3.6.3 (2020-02-29)
#Platform: x86_64-apple-darwin15.6.0 (64-bit)
#Running under: OS X  12.5.1

#Matrix products: default
#LAPACK: /Library/Frameworks/R.framework/Versions/
#3.6/Resources/lib/libRlapack.dylib

#locale:
#en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

#attached base packages:
#parallel  stats4 stats graphics  grDevices utils  datasets  methods   base

#other attached packages:
#RColorBrewer_1.1-3          pheatmap_1.0.12             apeglm_1.8.0
#DESeq2_1.26.0               SummarizedExperiment_1.16.1 DelayedArray_0.12.3
#BiocParallel_1.20.1         matrixStats_0.58.0          Biobase_2.46.0
#GenomicRanges_1.38.0        GenomeInfoDb_1.22.1         IRanges_2.20.2
#S4Vectors_0.24.4            BiocGenerics_0.32.0         forcats_0.5.1
#stringr_1.4.0               dplyr_1.0.5                 purrr_0.3.4
#readr_1.4.0                 tidyr_1.1.3                 tibble_3.1.6
#tidyverse_1.3.0             ggplot2_3.3.5

# code checked with lintr

```